File: src/test/java/no/nav/foreldrepenger/stønadskonto/regelmodell/konfig/ParameterTest.java
package no.nav.foreldrepenger.stønadskonto.regelmodell.konfig;


import static org.assertj.core.api.Assertions.assertThat;

import java.time.LocalDate;

import org.junit.jupiter.api.Test;

import no.nav.foreldrepenger.stønadskonto.regelmodell.grunnlag.Dekningsgrad;


class ParameterTest {

    @Test
    void periode_med_start_og_slutt_og_dato_utenfor_skal_ikke_overlappe() {
        var testPeriode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, LocalDate.of(2016, 1, 1), LocalDate.of(2018, 1, 1), 1);

        assertThat(testPeriode.overlapper(LocalDate.of(2019, 1, 1))).isFalse();
    }


    @Test
    void periode_uten_start_og_slutt_skal_overlappe() {
        var testPeriode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, null, null, 1);

        assertThat(testPeriode.overlapper(LocalDate.of(2017, 1, 1))).isTrue();
    }


    @Test
    void periode_med_start_og_slutt_skal_overlappe() {
        var testPeriode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, LocalDate.of(2016, 1, 1), LocalDate.of(2018, 1, 1), 1);

        assertThat(testPeriode.overlapper(LocalDate.of(2017, 1, 1))).isTrue();
    }

    @Test
    void periode_med_bare_start_skal_overlappe() {
        var testPeriode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, LocalDate.of(2016, 1, 1), null, 1);

        assertThat(testPeriode.overlapper(LocalDate.of(2017, 1, 1))).isTrue();
    }

    @Test
    void periode_med_bare_start_og_dato_før_start_skal_ikke_overlappe() {
        var testPeriode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, LocalDate.of(2016, 1, 1), null, 1);

        assertThat(testPeriode.overlapper(LocalDate.of(2015, 1, 1))).isFalse();
    }

    @Test
    void helePeriodenOverlapper() {
        var fom = LocalDate.now();
        var tom = fom.plusWeeks(2);
        var periode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom, tom, 1);

        assertThat(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom.minusDays(1), tom.plusDays(1), 1))).isTrue();
        assertThat(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom.plusDays(1), tom.minusDays(1), 1))).isTrue();
        assertThat(periode.overlapper(periode)).isTrue();
    }

    @Test
    void begynnelsenAvPeriodenOverlapper() {
        var fom = LocalDate.now();
        var tom = fom.plusWeeks(2);
        var periode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom, tom, 1);

        assertThat(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom.minusDays(1), tom.minusDays(1), 1))).isTrue();
        assertThat(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom, fom, 1))).isTrue();
        assertThat(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom.minusDays(1), fom, 1))).isTrue();
    }

    @Test
    void sluttenAvPeriodenOverlapper() {
        var fom = LocalDate.now();
        var tom = fom.plusWeeks(2);
        var periode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom, tom, 1);

        assertThat(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom.plusDays(1), tom.plusDays(1), 1))).isTrue();
        assertThat(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, tom, tom, 1))).isTrue();
        assertThat(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, tom, tom.plusDays(1), 1))).isTrue();
    }

    @Test
    void periodenRettFørOverlapperIkke() {
        var fom = LocalDate.now();
        var tom = fom.plusWeeks(2);
        var periode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom, tom, 1);

        assertThat(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom.minusDays(10), fom.minusDays(1), 1))).isFalse();
    }

    @Test
    void periodenRettEtterOverlapperIkke() {
        var fom = LocalDate.now();
        var tom = fom.plusWeeks(2);
        var periode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom, tom, 1);

        assertThat(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, tom.plusDays(1), tom.plusDays(5), 1))).isFalse();
    }


}

File: stonadskonto/src/konfig/Parameter.test.ts
import { describe, it, expect } from 'vitest';
import { Parameter } from './Parameter';
import { Dekningsgrad } from '../grunnlag/Dekningsgrad';

describe('Parameter', () => {
  it('periode_med_start_og_slutt_og_dato_utenfor_skal_ikke_overlappe', () => {
    const testPeriode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, new Date(2016, 0, 1), new Date(2018, 0, 1), 1);

    expect(testPeriode.overlappestDato(new Date(2019, 0, 1))).toBe(false);
  });

  it('periode_uten_start_og_slutt_skal_overlappe', () => {
    const testPeriode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, null, null, 1);

    expect(testPeriode.overlappestDato(new Date(2017, 0, 1))).toBe(true);
  });

  it('periode_med_start_og_slutt_skal_overlappe', () => {
    const testPeriode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, new Date(2016, 0, 1), new Date(2018, 0, 1), 1);

    expect(testPeriode.overlappestDato(new Date(2017, 0, 1))).toBe(true);
  });

  it('periode_med_bare_start_skal_overlappe', () => {
    const testPeriode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, new Date(2016, 0, 1), null, 1);

    expect(testPeriode.overlappestDato(new Date(2017, 0, 1))).toBe(true);
  });

  it('periode_med_bare_start_og_dato_før_start_skal_ikke_overlappe', () => {
    const testPeriode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, new Date(2016, 0, 1), null, 1);

    expect(testPeriode.overlappestDato(new Date(2015, 0, 1))).toBe(false);
  });

  it('helePeriodenOverlapper', () => {
    const fom = new Date();
    const tom = new Date(fom.getTime() + 2 * 7 * 24 * 60 * 60 * 1000);
    const periode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom, tom, 1);

    const before = new Date(fom.getTime() - 1 * 24 * 60 * 60 * 1000);
    const after = new Date(tom.getTime() + 1 * 24 * 60 * 60 * 1000);

    expect(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, before, after, 1))).toBe(true);

    const inside1 = new Date(fom.getTime() + 1 * 24 * 60 * 60 * 1000);
    const inside2 = new Date(tom.getTime() - 1 * 24 * 60 * 60 * 1000);
    expect(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, inside1, inside2, 1))).toBe(true);

    expect(periode.overlapper(periode)).toBe(true);
  });

  it('begynnelsenAvPeriodenOverlapper', () => {
    const fom = new Date();
    const tom = new Date(fom.getTime() + 2 * 7 * 24 * 60 * 60 * 1000);
    const periode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom, tom, 1);

    const before = new Date(fom.getTime() - 1 * 24 * 60 * 60 * 1000);
    const tomMinus = new Date(tom.getTime() - 1 * 24 * 60 * 60 * 1000);

    expect(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, before, tomMinus, 1))).toBe(true);
    expect(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom, fom, 1))).toBe(true);
    expect(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, before, fom, 1))).toBe(true);
  });

  it('sluttenAvPeriodenOverlapper', () => {
    const fom = new Date();
    const tom = new Date(fom.getTime() + 2 * 7 * 24 * 60 * 60 * 1000);
    const periode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom, tom, 1);

    const fomPlus = new Date(fom.getTime() + 1 * 24 * 60 * 60 * 1000);
    const tomPlus = new Date(tom.getTime() + 1 * 24 * 60 * 60 * 1000);

    expect(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fomPlus, tomPlus, 1))).toBe(true);
    expect(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, tom, tom, 1))).toBe(true);
    expect(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, tom, tomPlus, 1))).toBe(true);
  });

  it('periodenRettFørOverlapperIkke', () => {
    const fom = new Date();
    const tom = new Date(fom.getTime() + 2 * 7 * 24 * 60 * 60 * 1000);
    const periode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom, tom, 1);

    const before1 = new Date(fom.getTime() - 10 * 24 * 60 * 60 * 1000);
    const before2 = new Date(fom.getTime() - 1 * 24 * 60 * 60 * 1000);

    expect(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, before1, before2, 1))).toBe(false);
  });

  it('periodenRettEtterOverlapperIkke', () => {
    const fom = new Date();
    const tom = new Date(fom.getTime() + 2 * 7 * 24 * 60 * 60 * 1000);
    const periode = new Parameter(Dekningsgrad.DEKNINGSGRAD_100, fom, tom, 1);

    const after1 = new Date(tom.getTime() + 1 * 24 * 60 * 60 * 1000);
    const after2 = new Date(tom.getTime() + 5 * 24 * 60 * 60 * 1000);

    expect(periode.overlapper(new Parameter(Dekningsgrad.DEKNINGSGRAD_100, after1, after2, 1))).toBe(false);
  });
});

